### PR TITLE
Support internal cache

### PR DESCRIPTION
**Pull request by vchuravy**
replaces #534
adapts to JuliaLang/julia#52233

### CODE REVIEW

File: src/interface.jl [hunk 1]
```diff
@@ -176,10 +176,17 @@ runtime_module(@nospecialize(job::CompilerJob)) = error("Not implemented")
 isintrinsic(@nospecialize(job::CompilerJob), fn::String) = false
 
 # provide a specific interpreter to use.
+if isdefined(Base, :method_instance)
```
Comment (maleadt) [inline src/interface.jl:179]:
```suggestion
if VERSION >= v"1.11.0-DEV.1552"
```

I personally prefer a version-based check, especially here where `Base.method_instance` is not related to the `token` kwarg introduced here.

**Comment by maleadt** [comment]
Failure on 1.11:

```
julia> job, _ = Native.create_job(identity, (Int,))
(CompilerJob{NativeCompilerTarget, Main.Native.CompilerParams}(MethodInstance for identity(::Any), CompilerConfig for NativeCompilerTarget, 0x00000000000066fd), Base.Pairs{Symbol, Union{}, Tuple{}, @NamedTuple{}}())

julia> GPUCompiler.code_typed(job)
1-element Vector{Any}:
 CodeInfo(
1 ─     return x
) => Any

julia> Base.code_typed(identity, (Int,))
1-element Vector{Any}:
 CodeInfo(
1 ─     return x
) => Int64
```

**Comment by maleadt** [comment]
Could we also get some pretty printing back for the new cache? This is pretty sad:

```
julia> GPUCompiler.ci_cache(job)
Core.Compiler.WorldView{Core.Compiler.InternalCodeCache}(Core.Compiler.InternalCodeCache(GPUCompiler.GPUCompilerCacheToken(NativeCompilerTarget, false, # 0 methods for callable object)), Core.Compiler.WorldRange(0x00000000000066fd, 0x00000000000066fd))
```

**Comment by vchuravy** [comment]
> Could we also get some pretty printing back for the new cache? This is pretty sad:

Not really it is hard (maybe even impossible from Julia) to iterate over all cache entries. Since the entries are now a linked-list in `CodeInstance->cache`. In the Julia runtime we use a series of nested visitors over the global data-structures to find entries. 

So we would need something like `jl_foreach_reachable_mtable` 
`foreach_mtable_in_module` then followed by a  `jl_typemap_visitor` then visit all the method-instances and then look through all the code instances to print this. Feasible, but annoying. 

@vtjnash any other ideas?

**Comment by vchuravy** [comment]
> Failure on 1.11:

The issue is less sinister than I feared, but I am not sure jet what went wrong,

```
julia> job, _ = create_job(identity, (Int,))
(CompilerJob{NativeCompilerTarget, CompilerParams}(MethodInstance for identity(::Any), CompilerConfig for NativeCompilerTarget, 0x000000000000670b), Base.Pairs{Symbol, Union{}, Tuple{}, @NamedTuple{}}())

julia> job.source.specTypes
Tuple{typeof(identity), Any}
```

So we ended up with the wrong mi.

**Comment by vtjnash** [comment]
Yeah, I am not sure that pretty-printing a cache object is a sensible thing to try to enumerate

**Comment by maleadt** [comment]
With https://github.com/JuliaGPU/GPUCompiler.jl/pull/555 and https://github.com/JuliaLang/julia/pull/53723, this passes tests on 1.11, so we're getting close :-)

**Pull request merged**